util/threadpool_imp.cc

```diff
@@ -100,6 +102,46 @@ struct ThreadPoolImpl::Impl {
   // Set the thread priority.
   void SetThreadPriority(Env::Priority priority) { priority_ = priority; }
 
+  // The number of reserved threads could be fewer than the desired one
+  int ReserveThreads(int threads_to_be_reserved) {
+    std::unique_lock<std::mutex> lock(mu_);
+    int curr_reserved_threads =
+        reserved_threads_.load(std::memory_order_relaxed);
+    int reserved_threads_in_success = threads_to_be_reserved;
+    // Assume that at least one thread is left in threadpool. For example, when
+    // we use ReserveThreads in subcompaction, the thread that calls
+    // subcompaction should be taken into account in the threadpool
+    if (curr_reserved_threads + threads_to_be_reserved + 1 >
+        total_threads_limit_) {
```

Comment:
`total_threads_limit_` seems too high as some of those threads might be doing work and therefore should not be reserved. I think there should be a counter, `num_waiting_`, for threads in `bgsignal_.wait(lock)`. Then `num_waiting_` is the upper bound on how many threads can be reserved. Also you don't need to worry about `+ 1` because if this executes inside a thread-pool managed thread, that thread will naturally be excluded from `num_waiting_` so will be non-reservable.

Reply:
Indeed, `num_waiting_` is definitely a better upper bound than `total_threads_limit_`. I will add it in the next commit.